pilot/pkg/model/sidecar.go

```diff
@@ -371,7 +382,15 @@ func ConvertToSidecarScope(ps *PushContext, sidecarConfig *config.Config, config
 						sort.Strings(ns)
 						// Pick first namespace alphabetically
 						// This won't overwrite hostnames that have already been found eg because they were requested in hosts
-						addService(byNamespace[ns[0]])
+						var vss *Service
+						if matchPort {
+							vss = serviceMatchingListenerPort(byNamespace[ns[0]], listener)
```

Comment:
@ramaraochavali could you please kindly let me know if any design doc related to this change? Thanks.

we are using an old istio version, virtual service's destination is egress gateway service, which uses a port number different from sidecar's egress listener. This change cause it doesn't work -- serviceMatchingListenerPort() return nil.

Reply:
Unfortunately there is no design doc except for the description here https://github.com/istio/istio/pull/33113#issue-902198310.

Reply:
@ramaraochavali thanks for your reply. This breaks our egress gateway use case, so I trying to understand if any reason behind this change.

Reply:
The reason is the sidecar egress listener port is intended to restrict services that are explicitly listed in the sidecar definition. Later it has been enhanced to also bring in services that match the virtual service destinations - when we did that we did not consider port. So it was bringing destinations that does not have that matching port which are useless and is bringing lot more clusters. So that change was made.
I think there are two ways to solve this
- Either you explicitly list that service in your sidecar
- Ignore this check for external services
Not sure if there are other options @hzxuzhonghu any ideas?

Reply:
>Ignore this check for external services

I donot agree with this way, it makes external service being treated specially.

Not sure should we add some system namespace here, like [kube-system, istio-system], maybe not needed to

Reply:
Explicitly list that service

Reply:
> Either you explicitly list that service in your sidecar

@ramaraochavali do you mean add the wanted service in Sidecar's `egress` listener? that would cause extra outbound listener being generated, right?

Per my understanding, it's reasonable that route destination could be different from listener's port, so instead of checking destination service's port against listener port by
```
vss = serviceMatchingListenerPort(byNamespace[ns[0]], listener)
```
it should do
```
vss = serviceMatchingVirtualServicePorts(byNamespace[ns[0]], ports)
```

the `if matchPort` condition is not required here. Please kindly correct me if I misunderstand.

Reply:
matchPort block is used to look at **service** if egress listener has ports. As done [here](https://github.com/istio/istio/blob/8f7d578ce031dba8394104a90e5cfe31a9c417d8/pilot/pkg/model/sidecar.go#L360) for same namespace - We are doing the same here. Our goal is to always filter services based on port.

Reply:
Your goal is clear, but when inferring service from VirtualService destination, **the service should match route destination port, instead of listener's port**. Below if the latest code on master branch, IMO the `if matchPort` is not required:


```
					if len(ns) > 0 {
						sort.Strings(ns)
						// Pick first namespace alphabetically
						// This won't overwrite hostnames that have already been found eg because they were requested in hosts
						var vss *Service
						if matchPort {
							vss = serviceMatchingListenerPort(byNamespace[ns[0]], listener)
						} else {
							vss = serviceMatchingVirtualServicePorts(byNamespace[ns[0]], ports)
						}
						if vss != nil {
							addService(vss)
						}
					}

```